plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,849 @@
+#define DDC_VCP_CODE_SET_DATA_FF	0xF6
+#define DDC_VCP_CODE_COMMIT_FW   0xF7
+#define DDC_VCP_CODE_GET_ISP_MODE       0xF8
+#define DDC_VCP_CODE_REBOOT	 0xFB
+#define DDC_VCP_CODE_SUM	 0xFE
+#define DDC_VCP_CODE_VERSION	 0xFF
+#define FLASH_FIRMWARE_SIZE	 0x1E848
+#define DDC_DATA_FRAGEMENT_SIZE	 0x0B   /* 11 bytes for each DDC write */
+#define DDC_DATA_PAGE_SIZE		0x1000 /* 4K bytes for each block page */
+#define DDC_RW_MAX_RETRY_CNT     10
+
+/* timeout duration in ms to i2c-dev operation */
+#define FU_MEDIATEK_SCALER_DEVICE_IOCTL_TIMEOUT 5000
+
+/* delay time in ms before sending the next message */
+#define FU_MEDIATEK_SCALER_DDC_MSG_DELAY 50
+
+/* interval in ms between the poll to check device status */
+#define FU_MEDIATEK_SCALER_DEVICE_POLL_INTERVAL 1000
+
+typedef enum {
+	DDCCI_PRI_NORMAL,
+	DDCCI_PRI_UP,
+} FuMediatekScalerSetPriority;
+
+struct _FuMediatekScalerDevice {
+	FuUdevDevice parent_instance;
+	FuUdevDevice *i2c_dev;
+	gchar *i2c_dev_name;
+	gchar *dp_aux_dev_name;
+};
+
+G_DEFINE_TYPE(FuMediatekScalerDevice, fu_mediatek_scaler_device, FU_TYPE_UDEV_DEVICE)
+
+static gboolean
+fu_mediatek_scaler_ensure_device_address(FuMediatekScalerDevice *self,
+					 guint8 address,
+					 GError **error)
+{
+	g_autoptr(GError) error_local = NULL;
+
+	if (!fu_udev_device_ioctl(self->i2c_dev,
+				  I2C_SLAVE,
+				  (guint8 *)(guintptr)address,
+				  NULL,
+				  FU_MEDIATEK_SCALER_DEVICE_IOCTL_TIMEOUT,
+				  &error_local)) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "Failed to set address on '%s' for '%s': %s",
+			    self->i2c_dev_name,
+			    self->dp_aux_dev_name,
+			    (error_local != NULL ? error_local->message : "unknown"));
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_set_i2c_dev(FuMediatekScalerDevice *self,
+				      const GPtrArray *i2c_devices,
+				      GError **error)
+{
+	for (guint i = 0; i < i2c_devices->len; i++) {
+		FuUdevDevice *i2c_device = g_ptr_array_index(i2c_devices, i);
+		g_autoptr(GPtrArray) i2c_devs =
+		    fu_udev_device_get_children_with_subsystem(i2c_device, "i2c-dev");
+
+		if (i2c_devs->len == 0) {
+			g_debug("no i2c-dev found under %s",
+				fu_udev_device_get_sysfs_path(i2c_device));
+			continue;
+		}
+		if (i2c_devs->len > 1) {
+			g_debug("ignoring %u additional i2c-dev under %s",
+				i2c_devs->len - 1,
+				fu_udev_device_get_sysfs_path(i2c_device));
+		}
+
+		/* the first i2c_dev is enforced to represent the dp aux device */
+		self->i2c_dev = g_object_ref(g_ptr_array_index(i2c_devs, 0));
+		self->i2c_dev_name =
+		    g_strdup(fu_udev_device_get_device_file(FU_UDEV_DEVICE(self->i2c_dev)));
+		g_debug("found I2C bus at %s, using this device",
+			fu_udev_device_get_sysfs_path(self->i2c_dev));
+
+		/* the following open() requires the device have IDs set */
+		if (!fu_udev_device_set_physical_id(self->i2c_dev, "i2c", error))
+			return FALSE;
+
+		return self->i2c_dev != NULL;
```

Comment:
I don't think it's possible for this to be `NULL`,  you set it on line 106.  you should just `return TRUE;`

Reply:
Furthermore if it was possible - you need to set error in that case.

Reply:
Fixed